docs/architecture/adr-022-custom-panic-handling.md

```diff
@@ -0,0 +1,208 @@
+# ADR 022: Custom BaseApp panic handling
+
+## Changelog
+
+- 2020 Apr 24: Initial Draft
+
+## Context
+
+The current implementation of BaseApp does not allow developers to write custom error handlers during panic recovery
+[runTx()](https://github.com/cosmos/cosmos-sdk/blob/bad4ca75f58b182f600396ca350ad844c18fc80b/baseapp/baseapp.go#L539)
+method. We think that this method can be more flexible and can give SDK users more options for customizations without
+the need to rewrite whole BaseApp. Also there's one special case for `sdk.ErrorOutOfGas` error handling, that case
+might be handled in a "standard" way (middleware) alongside the others.
+
+We propose middleware-solution, which could help developers implement the following cases:
+* add external logging (let's say sending reports to external services like [Sentry](https://sentry.io));
```

Comment:
I do not understand how this would work. Would every node make an HTTP call to Sentry?

Reply:
Those are examples of how developers can use those middlewares. They can send panic logs to Sentry if they need to or do other struf needed.

Reply:
I still don't understand how Sentry would be used in a replicated state machine.

Reply:
Those are just examples of what developers of Cosmos SDK based project could use those handlers for. May be they want to get notifications on every `panic()` event (Sentry), that doesn't collide with state machine ideology.

Reply:
The issue is that (in a production blockchain) the state machine is *replicated*, so many copies of every error would be sent to Sentry, and you would need to be careful to ensure that making HTTP requests doesn't introduce non-determinism. It's an alright example, but these points need to be clarified.

Reply:
Actually Sentry can aggregate multiple similar event into one issue with some meta and statistics. As for non-deterministic behavior: developers can do that in so many ways including the proposed middlewares =)

Reply:
That's nice. The determinism bit is more important. I think it would be sufficient if we add a warning that middlewares must be deterministic, or consensus safety will be lost.